### PR TITLE
Document and automate the process to fetch from upstream

### PR DESCRIPTION
**Pull request by joshpeterson**
The process of fetching changes from the upstream dotnet/runtime
repository will be done weekly by a Yamato job. This job will

1. Fetch from the `main` branch upstream
2. Create a branch of `unity-main` named with the current date
3. Open a pull request to merge the new branch into `unity-main`

### CODE REVIEW

**Comment by joshpeterson** [comment]
You can see an example of the automatically generated pull request here: https://github.com/Unity-Technologies/runtime/pull/7

File: .yamato/update_from_upstream.yml
Comment (joncham) [inline .yamato/update_from_upstream.yml:12]: This indicates merge commits will use your user id?

File: .yamato/scripts/update_from_upstream.sh
Comment (joncham) [inline .yamato/scripts/update_from_upstream.sh:8]: You could just get `main` I think, 
```
git fetch upstream main
```

File: .yamato/scripts/update_from_upstream.sh
Comment (joncham) [inline .yamato/scripts/update_from_upstream.sh:10]: would it be better to
```
git merge -ff-only upstream/main
```

I want to ensure we fail if a fast forward is not possible.

File: .yamato/scripts/update_from_upstream.sh
Comment (joncham) [inline .yamato/scripts/update_from_upstream.sh:7]: Do we need error checking for all these commands?

File: .yamato/scripts/update_from_upstream.sh
Comment (joncham) [inline .yamato/scripts/update_from_upstream.sh:18]: Should we pass `--utc`?

File: .yamato/scripts/update_from_upstream.sh
Comment (joshpeterson) [reply .yamato/scripts/update_from_upstream.sh:7]: Good point - I think we need to add `set -e` at the top of this file to get that behavior.

File: .yamato/scripts/update_from_upstream.sh
Comment (joshpeterson) [reply .yamato/scripts/update_from_upstream.sh:8]: I'll try this, thanks!

File: .yamato/scripts/update_from_upstream.sh
Comment (joshpeterson) [reply .yamato/scripts/update_from_upstream.sh:10]: Yes, that sounds like a good idea.

File: .yamato/scripts/update_from_upstream.sh
Comment (joshpeterson) [reply .yamato/scripts/update_from_upstream.sh:18]: I don't think it matters too much, so long as we only do one PR per day.

File: .yamato/update_from_upstream.yml
Comment (joshpeterson) [reply .yamato/update_from_upstream.yml:12]: Yes - I'm not sure that we have a better option. I decided to do this for now.

File: .yamato/scripts/update_from_upstream.sh
Comment (joshpeterson) [inline .yamato/scripts/update_from_upstream.sh:12]: @joncham It looks like a fast forward merge is failing with the error:

```
fatal: Not possible to fast-forward, aborting.
```

So maybe we should keep this as a rebase. That way our `main` always has the same history as upstream's `main`.

File: .yamato/scripts/update_from_upstream.sh
Comment (joshpeterson) [inline .yamato/scripts/update_from_upstream.sh]: @joncham It looks like we need the previous set of Git commands here, otherwise we will end up in a detached head state that won't work later.

**Comment by joshpeterson** [comment]
I think we have the scripts in a good state now. The latest test PR is https://github.com/Unity-Technologies/runtime/pull/9.

If these make sense, I'll squash the commits down before merging. Let me know!

**Review by UnityAlex — APPROVED** [review]

File: .yamato/scripts/update_from_upstream.sh
Comment (joncham) [reply .yamato/scripts/update_from_upstream.sh]: If we cannot fast forward that indicates a problem AFAIK. It means either we commited to `main` (wrong) or upstream rewrote history?

Comment (joshpeterson) [reply .yamato/scripts/update_from_upstream.sh:13]: Ahh, you are correct. I was not checking out `main`, the clone was still on `unity-main`, which is why this failed. With the latest code the fast forward merge works.

**Comment by joshpeterson** [comment]
This is working now with a fast-forward merge on `main` rather than a rebase.

**Comment by joshpeterson** [comment]
I've squashed the changes down now into their final form, I hope!

File: .yamato/scripts/update_from_upstream.sh
Comment (scott-ferguson-unity) [inline .yamato/scripts/update_from_upstream.sh:16]: Should we have the `-f` here?  If we're `--ff-only` we shouldn't have to force push right?  And if `upstream/main` doesn't push cleanly we'd probably want the `push` to fail.

Comment (joshpeterson) [reply .yamato/scripts/update_from_upstream.sh:16]: Good point, we can remove `-f` here. I had that earlier due to the rebase. But we don't want that any more.

Comment (joshpeterson) [reply .yamato/scripts/update_from_upstream.sh]: Fixed now!

File: .yamato/scripts/update_from_upstream.sh
Comment (scott-ferguson-unity) [inline .yamato/scripts/update_from_upstream.sh:27]: Do we need to create a PR here?  We'd already fail at https://github.com/Unity-Technologies/runtime/pull/8/files#diff-ef7c119a3b7f6122d886f3c65142eafe9dd9fd0edaadbfeeb675490477aab298R23 if the merge fails.  Could we run the tests as part of the yamato job and merge the results automatically if it passes?  Or do we want the manual review of the changes?

**Review by scott-ferguson-unity — APPROVED** [review]

File: .yamato/scripts/update_from_upstream.sh
Comment (joshpeterson) [reply .yamato/scripts/update_from_upstream.sh:27]: > Could we run the tests as part of the yamato job and merge the results automatically if it passes?

This might be possible. I could see:

1. Yamato job to do the fetch from upstream
2. Yamato job to run the tests
3. Yamato job to merge the branch into unity-main

But this seemed like a simpler workflow.

> Or do we want the manual review of the changes?

I kind of did want a manual review, but looking at the test pull requests, they are probably too big to review.

My preference is to do this simpler set up now. If the pull request review/merge becomes rote, then we can change the process have the machines do it for us later.

**Review by brian-taylor-unity — APPROVED** [review]

File: .yamato/update_from_upstream.yml
Comment (joncham) [inline .yamato/update_from_upstream.yml:5]: I'm not sure it matters from resource point of view, but do we need xlarge?

**Review by joncham — APPROVED** [review]

File: .yamato/update_from_upstream.yml
Comment (joshpeterson) [reply .yamato/update_from_upstream.yml:5]: I'm not sure about that - I copied what some other jobs in IL2CPP used. I say we stick with this for now and see how things go in the future.

File: unity/README.md
Comment (bholmes) [inline unity/README.md]: This link is wrong?  https://github.cds.internal.unity3d.com/unity/il2cpp/pull/3602

File: unity/README.md
Comment (bholmes) [inline unity/README.md]: I suspect we mean unity-main here and not main? And why do this at all?  The upstream merge should get this later. Or conflict? I may just be confused, which points to more clarity in this doc is needed.

**Review by bholmes — CHANGES_REQUESTED** [review]

File: unity/README.md
Comment (joshpeterson) [reply unity/README.md]: Yes! I'll correct this.

File: unity/README.md
Comment (joshpeterson) [reply unity/README.md]: Good point, maybe this is not necessary. I thought was that we don't want to wait to merge changes in our fork for them to first land upstream, then wait a week for the sync to happen from upstream.

My idea here is that we merge the changes to `unity-main` in our fork, then create a new branch for the upstream pull request. So long as we cherry-pick the commits from `unity-main`, the sync after they land upstream should work cleanly.

File: unity/README.md
Comment (joshpeterson) [reply unity/README.md]: Fixed

File: unity/README.md
Comment (bholmes) [reply unity/README.md]: I still do not think you want `main` in this section. We should never to PR into Unity's main.  Only Unity's `unity-main`. I think that our `main` should never have any different commits from upstream.

Also consider deleting our `main`.  We do not need it.

Comment (joshpeterson) [reply unity/README.md]: > We should never to PR into Unity's main.

I agree. But don't we need to make a branch of the Unity fork of `main` to allow us to open a PR to the upstream `main`?

> Also consider deleting our main. We do not need it.

I think we do need a branch to match the upstream `main`, right?

**Comment by joshpeterson** [comment]
After some discussion with @bholmes, it turns out that he is correct - we don't need to use `main` in this fork. I've updated the scripts, the docs, and the PR description to use this approach, which is simpler. Please re-review these changes.

File: unity/README.md
Comment (bholmes) [inline unity/README.md:9]: Should we add a link to the Yamato job? (We have a job, just not telling you where it is.  :) )

**Review by bholmes — APPROVED** [review]

File: unity/README.md
Comment (joshpeterson) [reply unity/README.md:9]: I was thinking about that as well. I did not add it because the Yamato jobs are not accessible publicly.

Comment (bholmes) [reply unity/README.md:9]: I did forget this was public.

**Pull request merged**
